lib/pgBackRest/Storage/Local.pm

```diff
@@ -0,0 +1,775 @@
+    }
+
+    # Return from function and log return values if any
+    return logDebugReturn
+    (
+        $strOperation,
+        {name => 'self', value => $self}
+    );
+}
+
+####################################################################################################################################
+# pathGet
+####################################################################################################################################
+sub pathGet
+{
+    my $self = shift;
+
+    # Assign function parameters, defaults, and log debug info
+    my
+    (
+        $strOperation,
+        $strPathExp,                                               # File that that needs to be translated to a path
+        $bTemp,                                                    # Return the temp file name
+    ) =
+        logDebugParam
+    (
+        __PACKAGE__ . '->pathGet', \@_,
+        {name => 'strPathExp', required => false, trace => true},
+        {name => 'bTemp', optional => true, default => false, trace => true},
+    );
+
+    # Path and file to be returned
+    my $strPath;
+    my $strFile;
+
+    # Is this an absolute path type?
+    my $bAbsolute = false;
+
+    if (defined($strPathExp) && index($strPathExp, qw(/)) == 0)
+    {
+        $bAbsolute = true;
+        $strPath = $strPathExp;
+    }
+    else
+    {
+        # Is it a rule type
+        if (defined($strPathExp) && index($strPathExp, qw(<)) == 0)
+        {
+            # Extract the rule type
+            my $iPos = index($strPathExp, qw(>));
+
+            if ($iPos == -1)
+            {
+                confess &log(ASSERT, "found < but not > in '${strPathExp}'");
+            }
+
+            my $strType = substr($strPathExp, 0, $iPos + 1);
+
+            # Extract the filename
+            if ($iPos < length($strPathExp) - 1)
+            {
+                $strFile = substr($strPathExp, $iPos + 2);
+            }
+
+            # Lookup the rule
+            if (!defined($self->{hRule}->{$strType}))
+            {
+                confess &log(ASSERT, "storage rule '${strType}' does not exist");
+            }
+
+            # If rule is a ref then call the function
+            if (ref($self->{hRule}->{$strType}))
+            {
+                $strPath = $self->{strPathBase};
+                $strFile = $self->{hRule}{$strType}{fnRule}->($strType, $strFile, $self->{hRule}{$strType}{xData});
+            }
+            # Else get the path
+            else
+            {
+                $strPath = $self->pathBase() . "/$self->{hRule}->{$strType}";
+            }
+        }
+        # Else it must be relative
+        else
+        {
+            $strPath = $self->{strPathBase};
+            $strFile = $strPathExp;
+        }
+    }
+
+    # Make sure a temp file is valid for this type and file
+    if ($bTemp)
+    {
+        # Only allow temp files for PATH_REPO_ARCHIVE, PATH_SPOOL_ARCHIVE_OUT, PATH_REPO_BACKUP_TMP and any absolute path
```

Comment:
CSHANG Is this comment correct?  These constants don't exist - the PATH has been replaced with STORAGE - even so, how do we know if this boolean is set correctly if we're not checking those settings? And if we determined it was an absolute path, does that mean bAllowTemp should always be true?